data_loader.py

```diff
@@ -28,45 +31,103 @@ def disconnect(self):
+        """Sufficent flattening dict"""
+        if type(data) == dict and len(data) != 0:
+            keystring = keystring + "." if keystring else keystring
+            for k in data:
+                yield from self._flatten_dict(data[k], keystring + str(k))
+        else:
+            yield keystring, data
+
+    def _combine_excel(self, list_of_articles: list):
+        """Add excel values to all articles"""
+        review_status = pd.read_excel(
+            self.args.reference_data_file,
+            sheet_name="review_status",
+            header=2,
+            usecols="B:E",
+            dtype=str,
+        )
+        date_completed = pd.read_excel(
+            self.args.reference_data_file, sheet_name="date_completed"
+        )
+
+        # Preprocessing data
+        review_status.columns = review_status.columns.str.replace(" ", "_").str.lower()
+        date_completed.columns = date_completed.columns.str.replace(
+            " ", "_"
+        ).str.lower()
+        review_status["article_id"] = review_status["article_id"].apply(
+            lambda x: x.strip()
+        )
+        date_completed["reference_id"] = date_completed["reference_id"].values.astype(
+            str
+        )
+
+        for article in list_of_articles:
+            excel_dict = {}
+            article_status = review_status.loc[
+                review_status["article_id"] == article["_id"]
+            ]
+            if len(article_status) > 1:
+                excel_dict["status.duplicates"] = len(article_status)
+                excel_dict["status.duplicates.reference_id"] = article_status[
+                    "reference_id"
+                ].tolist()
+
+            if article_status.empty:
+                excel_dict.update(article_status.to_dict())
+                reference_id = None
+            else:
+                excel_dict.update(article_status.tail(1).to_dict("records")[0])
+                reference_id = article_status["reference_id"].values
+
+            article_date = date_completed.loc[
+                date_completed["reference_id"].isin(reference_id)
+            ]
+
+            if article_date.empty:
+                excel_dict.update(article_date.to_dict())
+            else:
+                excel_dict.update(article_date.tail(1).to_dict("records")[0])
+
+            article.update(excel_dict)
+
+        return list_of_articles
+
     def getDataBatch(self, batch_size):
         """
         Generator - Get data from source on batches.
 
         :returns One list for each batch. Each of those is a list of
                  dictionaries with the defined rows.
         """
-        # TODO: implement - this dummy implementation returns one batch of data
-        yield [
-            {
-                "headline.main": "The main headline",
-                "_id": "1234",
-            }
+        with open(self.args.api_response_file) as json_file:
+            data = json.load(json_file)
```

Comment:
Personally, I would extract it to a separate function like `get_api_data` or whatsoever :)